### DISCUSSION TITLE
help with TypedResult.read please

### DISCUSSION THREAD

**Discussion by matthewkooshad** [simolus3/drift#1231] [Q&A]
I just updated my version of moor. the result.read below is no longer working for me, and I couldn't figure out how to provide an expression to it to reference the fields from table1. if there's a cleaner way to manage doing this join query and reading some fields into List<MyCustomClass>, please let me know also. thanks!

```dart
final query = select(table3).join([
      innerJoin(table2, table2.id.equalsExp(table3.id), useColumns: false),
      innerJoin(table1, table1.id.equalsExp(table2.table1Id), useColumns: false),
    ]);

    query.addColumns([table1.title]);

final results = await query.get();

    return results.map((result) {
      //print(result.rawData.data);
      return MyCustomClass(
        prop1: result.readTable(table3),
        prop2: Table1PartialClass(
          id: result.read(table1.id),
          title: result.read(table1.title),
        ),
      );
    }).toList();
```

**Comment by simolus3** [comment]
Can you describe which columns you're actually interested in? You're setting `useColumns: false` for `table1` and then add `table1.title` later (this is fine). But then, you use `table1.id` in `map`, I wouldn't expect that column to be available based on the query you posted.

If you only need a subset of `table1`'s columns, joining it with `useColumns: false` and later adding the columns seems fine to me. Perhaps you just meant to use `addColumns([table1.id, table1.title])`?

**Reply by matthewkooshad** [reply]
i can gladly try that. but, i think the problem is with what i'm providing to the read method. it worked fine in previous version i was using of moor.

what should these lines below look like?
id: result.read(table1.id),
title: result.read(table1.title),

it is looking for an Expression it seems. how would i provide an expression to that read method to obtain the value for those fields in the read?

**Comment by matthewkooshad** [comment]
the exact error for those read method calls:
The argument type 'GeneratedIntColumn' can't be assigned to the parameter type 'Expression<int>'.dartargument_type_not_assignable
The argument type 'GeneratedTextColumn' can't be assigned to the parameter type 'Expression<String>'.dartargument_type_not_assignable

**Reply by simolus3** [reply]
That is very weird - `GeneratedIntColumn` definitely inherits from `Expression`! Can you try running `flutter clean` and see if that helps at all?

**Reply by matthewkooshad** [reply]
i ran flutter clean just now. then, i ran flutter pub get. then, i ran flutter pub run build_runner watch --delete-conflicting-outputs , which was successful; waiting with this as the last response:
[INFO] Succeeded after 31.4s with 14 outputs (197 actions)

among my repo, i only have these errors. no other files after errors. and, i had many errors before this, as i just updated this:
sdk: ">=2.12.0 <3.0.0"

so, i had to take care of all of the null safety related stuff, which before working on those problems, i also updated:
dependencies:
...
  moor_db_viewer: ^4.0.0 #https://pub.dev/packages/moor_db_viewer
  moor_flutter: ^4.0.0 #https://pub.dev/packages/moor_flutter

dev_dependencies:
  build_runner:
  moor_generator: ^4.3.1 #https://pub.dev/packages/moor_generator

here's my flutter --version output:
Flutter 2.2.0-10.2.pre • channel beta •
https://github.com/flutter/flutter.git
Framework • revision b5017bf8de (4 weeks ago) •
2021-04-28 17:09:53 -0700
Engine • revision 91ed51e05c
Tools • Dart 2.13.0 (build 2.13.0-211.13.beta)

**Reply by matthewkooshad** [reply]
here's some info from the file where those 2 line errors i've asked for help on:
import 'package:moor/moor.dart';
import 'package:moor_flutter/moor_flutter.dart';
also have imports for db files that have the 3 tables. all of the other code in the Dao for these tables is fine. i've changed the table names to simple ones with numbers to not share out any app domain info.

@UseDao(tables: [Table3s, Table2s, Table1s])
class Table3sDao extends DatabaseAccessor<Database> with _$Table3sDaoMixin {

**Reply by matthewkooshad** [reply]
the read method on my local is coming from:
flutter\.pub-cache\hosted\pub.dartlang.org\moor-4.3.1\lib\src\runtime\query_builder\statements\select\select.dart

**Reply by simolus3** [reply]
If you go to `lib\src\dsl\columns.dart` you'll see that `Column` does extend `Expression. I don't think there are any errors in your code either.
If you open the generated code, do you get analysis errors in there as well?

**Reply by matthewkooshad** [reply]
yes, i normally exclude .g files from the analyzer, but i commented out just now the below line in my analysis_options.yaml:
exclude:
    - lib/db/*.g.dart

now, i see errors there too for database.g.dart but, it's not the same error or same looking code with the error in the database.g.dart. it is only complaining there about `GeneratedDatabase db` with each of my db tables: 
Avoid defining unused parameters in constructors.dartavoid_unused_constructor_parameters

**Reply by matthewkooshad** [reply]
just to confirm, yes, in flutter\.pub-cache\hosted\pub.dartlang.org\moor-4.3.1\lib\src\dsl\columns.dart , i do see 
abstract class Column<T> extends Expression<T> {
and the diff data type columns extend Column

since the code looks good, any thing else i can pass into that Read method or another approach on just getting certain columns selected from the join query?

**Reply by simolus3** [reply]
>  Avoid defining unused parameters in constructors.dartavoid_unused_constructor_parameters

Yeah that should be fine, unfortunately we can't stop generating the now unused parameter until the next breaking moor release. Excluding generated files by default is also the right call, I was just wondering if we were perhaps missing any imports.

Can you try closing all Dart editors/IDEs and then delete `%LOCALAPPDATA%\.dartServer` (make a backup first just to be sure)? That's kind of a last resort when the analyzer is being weird, I really don't see an obvious error here.

**Reply by matthewkooshad** [reply]
i did those 2 steps you've last mentioned. after i reopened vs code, and my repo folder in vs code, it initially showed no problems. but, then the analyzer ran, and i saw the .dartServer folder reappear in the explorer window i still had open. then, the exact same problems showed in vs code, nothing more, nothing less. i hope you have some other ideas. 

if no other ideas, as a final resort, i think maybe you have some idea on how i could read those fields in another way from the join query.

**Reply by simolus3** [reply]
Can you reproduce this with a smaller project or privately send me the problematic files at oss@simonbinder.eu? We have tests in `moor` using `.read()` with columns, so I don't understand how there could be an issue from the analyzer. Maybe there's something else we're overlooking here.

**Reply by matthewkooshad** [reply]
thanks, i'll plan to reproduce it outside of my app and let you know if i'm able to do so, as well as share the files to you if so. thanks for your help!
